docs/_posts/2022-05-06-writing-cli-tools-with-typescript.md

```diff
@@ -0,0 +1,187 @@
+
+Nonetheless, a typically non-trivial part of any TypeScript project is the set up. Writing a CLI with e.g. Python and the [Click](https://click.palletsprojects.com/en/8.1.x/) library requires little more than writing a file with the [main function](https://docs.python.org/3/library/__main__.html#packaging-considerations). The higher setup complexity of TypeScript is of course due to the nature of TypeScript being a superset of JavaScript. To run TypeScript code, you typically transpile it to JavaScript and then execute the JavaScript code with Node.js. (Yes, even [ts-node](https://github.com/TypeStrong/ts-node) does that.) So, before running a program written in TypeScript, we need to transpile it to JavaScript first. This requires tooling and therefore setup.
+
+Preferably, we also want to have a single, packaged binary as the result of building our project, rather than a script file, to reduce installation overhead on part of the user. 
+
+To assure high code quality we also typically want to include code formatting, linting, and tests.
+
+Preferably, we also want to have a single, packaged binary as the result of building our project rather than a script file to reduce installation overhead on part of the user. 
+
+To assure high code quality we also typically want to include code formatting, linting, and tests.
+
+## Basic development flow
+
+To hit the ground running, I created a starter project which already contains all required boilerplate and dependencies for running, formatting, linting, testing the code and packaging it to a binary. 
+
+> ℹ️ The starter project can be found on my [GitHub](https://github.com/fkurz/typescript-cliutil-starter) and via [npm](https://www.npmjs.com/package/@fkurz/typescript-cliutil-starter/). 
+
+```bash
+git clone git@github.com:fkurz/typescript-cliutil-starter.git my-cli
+cd my-cli 
+npm i
+```
+
+Our starter project uses 
+
+- the [*commander*](https://www.npmjs.com/package/commander) package for argument parsing and executing commands, 
+- code linting and formatting with [eslint](https://eslint.org/) and [prettier](https://prettier.io/), 
+- [Jest](https://jestjs.io/) as test runner and assertion library, and 
+- [pkg](https://github.com/vercel/pkg) to build a stand-alone binary.
+
+To illustrate the development flow, we now add an example subcommand and test it. 
+
+For example, we could add a command *yell* which is exactly like the *say* command (that already exists in the project) but a bit less subtle.
+
+To do this, we add a file *src/yell.ts* and then add the following code.
+
+```typescript
+// src/yell.ts
+import { Command } from "commander";
+
+export default (): Command =>
+ new Command()
+  .name("yell")
+  .description("Say the word passed as the first argument but louder")
+  .argument("<word>")
+  .action((word: string) => console.log(word.toLocaleUpperCase()));
+```
+
+
+
```

Comment:
Markdown rendering doesn't care about the number of empty lines. They count as one and simply create a new paragraph. I suppose this is just for structuring the source markdown file?

Reply:
Unintentional.